# Lending Case Study
Lending Club Loan dataset contains information about past loan applicants and whether they are defaulted or not. The aim is to identify patterns that indicates if a person is likely to default. Based on this problem statement, data is anlysed and cleansed.

**Data Quality Issues in Loan Dataset**
| Issue                     | Description                                   | Example                               |
|---------------------------|-----------------------------------------------|---------------------------------------|
| Missing values            | Many columns contain missing values           | mths_since_last_record, tax_liens     |
| Data entry errors         | Some column data have spelling mistakes       | home_ownership, purpose               |
| Non-standardised format   | Some columns have numerical values with units | term, int_rate, emp_length, zip_code  |
| Irrelevant columns        | Some columns are too descriptive in native    | url, desc, title                      |
| Same valued columns       | Some columns have the same data               | payment_loan, policy_code             |
| Unique identifier columns | Some columns are unique identifiers           | id, member_id                         |

### Import Libraries

In [100]:
import pandas as pd

### Data Load

In [ ]:
df = pd.read_csv('cleansed_loan.csv', na_values=['NA', 'n/a'])
df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,...,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,open_acc,pub_rec,revol_bal,revol_util,total_acc,pub_rec_bankruptcies
0,5000,5000,4975.0,36 months,10.65%,162.87,B,B2,10+ years,RENT,...,0,Jan-85,1,NaN,3,0,13648,83.70%,9,0.0
1,2500,2500,2500.0,60 months,15.27%,59.83,C,C4,< 1 year,RENT,...,0,Apr-99,5,NaN,3,0,1687,9.40%,4,0.0
2,2400,2400,2400.0,36 months,15.96%,84.33,C,C5,10+ years,RENT,...,0,Nov-01,2,NaN,2,0,2956,98.50%,10,0.0
3,10000,10000,10000.0,36 months,13.49%,339.31,C,C1,10+ years,RENT,...,0,Feb-96,1,35.0,10,0,5598,21%,37,0.0
4,3000,3000,3000.0,60 months,12.69%,67.79,B,B5,1 year,RENT,...,0,Jan-96,0,38.0,15,0,27783,53.90%,38,0.0


# Data Cleansing

### Fixing Rows and Columns

**Fixing Rows**
- No summary rows.
- No incorrect header or footer rows.
- No extra rows like column number, indicators, blank rows, page no.

**Fixing Columns**
- Merging columns is not necessary since each column has its own purpose
- Column data transformation
    - `term` column contains values like '36 months'. This can be split to have only numeric values.
    - `int_rate` column also contains values like '10.65%'. This can be split to have only numeric values for easier calcultion.
    - Similarly `emp_length` and `zip_code` can also be standardised.
- No column names are missing.
- Column rename is not necessary since each column name has a meaning as described in the Data Dictionary.
- No misaligned columns

## Filter Data

- Removed the below unnecessary columns.
- Row removal and data aggregation might not be required.

    | Column Name                        | Reason for Deletion                  |
    |------------------------------------|--------------------------------------|
    | url, desc, title                   | Irrelevant/free text columns         |
    | application_type                   | often only one value                 |
    | emp_title                          | Free text                            |
    | payment_plan                       | Always 'n'                           |
    | mths_since_last_record             | Too many missing values              |
    | initial_list_status                | Only one value ('f' - franctional)   |
    | out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int | Post-loan metrics, NA |    
    | policy_code                        | Always '1'                           |
    | next_pymnt_d, last_credit_pull_d, last_pymnt_d, last_pymnt_amnt, total_rec_late_fee, collection_recovery_fee, recoveries | Post loan data, NA                    |
    | id, member_id                      | Unique identifiers                   |
    | total_il_high_credit_limit, total_bc_limit, total_bal_ex_mort, tot_hi_cred_lim, tax_liens, percent_bc_gt_75, pct_tl_nvr_dlq, num_tl_op_past_12m, num_tl_90g_dpd_24m, num_tl_30dpd, num_tl_120dpd_2m, num_sats, num_rev_tl_bal_gt_0, num_rev_accts, num_op_rev_tl, num_il_tl, num_bc_tl, num_bc_sats, num_actv_rev_tl, num_actv_bc_tl, num_accts_ever_120_pd, mths_since_recent_revol_delinq, mths_since_recent_inq, mths_since_recent_bc_dlq, mths_since_recent_bc, mort_acc, mo_sin_rcnt_tl, mo_sin_rcnt_rev_tl_op, mo_sin_old_rev_tl_op, mo_sin_old_il_acct, delinq_amnt, chargeoff_within_12_mths, bc_util, bc_open_to_buy, avg_cur_bal, acc_open_past_24mths, inq_last_12m, total_cu_tl, inq_fi, total_rev_hi_lim, all_util, max_bal_bc, open_rv_24m, open_rv_12m, il_util, total_bal_il, mths_since_rcnt_il, open_il_24m, open_il_12m, open_il_6m, open_acc_6m, tot_cur_bal, tot_coll_amt, acc_now_delinq, verification_status_joint, dti_joint, annual_inc_joint, mths_since_last_major_derog, collections_12_mths_ex_med | Columns have no values ('NA' or '0')          |
    

## Missing values
- Some columns such as mths_since_last_record have too many missing data, can be removed.
- Some columns have some data missing with annotations 'NA' or 'n/a' or empty. These are converted to numerical values.
- Standardise precision for float columns such as term, int_rate.
- Columns are scaled correctly.
- Removed extra characters such as units and 'xx' for zip_code.
- All text columns are standardised to lower case.

### Find columns with missing values

In [73]:
missingColumnValues = df.isnull().sum()
missingColumnValues

loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
emp_length                 1075
home_ownership                0
annual_inc                    0
verification_status           0
issue_d                       0
loan_status                   0
purpose                       0
zip_code                      0
addr_state                    0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
mths_since_last_delinq    25682
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                   50
total_acc                     0
pub_rec_bankruptcies        697
dtype: int64

### Handle missing values for columns 

In [ ]:
df['emp_length'] = df['emp_length'].fillna(0)
df['mths_since_last_delinq'] = df['mths_since_last_delinq'].fillna(0)
df['revol_util'] = df['revol_util'].fillna(0)
df['pub_rec_bankruptcies'] = df['pub_rec_bankruptcies'].fillna(0)

df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,...,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,open_acc,pub_rec,revol_bal,revol_util,total_acc,pub_rec_bankruptcies
0,5000,5000,4975.0,36 months,10.65%,162.87,B,B2,10+ years,RENT,...,0,Jan-85,1,0.0,3,0,13648,83.70%,9,0.0
1,2500,2500,2500.0,60 months,15.27%,59.83,C,C4,< 1 year,RENT,...,0,Apr-99,5,0.0,3,0,1687,9.40%,4,0.0
2,2400,2400,2400.0,36 months,15.96%,84.33,C,C5,10+ years,RENT,...,0,Nov-01,2,0.0,2,0,2956,98.50%,10,0.0
3,10000,10000,10000.0,36 months,13.49%,339.31,C,C1,10+ years,RENT,...,0,Feb-96,1,35.0,10,0,5598,21%,37,0.0
4,3000,3000,3000.0,60 months,12.69%,67.79,B,B5,1 year,RENT,...,0,Jan-96,0,38.0,15,0,27783,53.90%,38,0.0


### Standardize the format of some columns and convert them to numeric type

In [ ]:
df['term'] = df['term'].astype(str).str.replace(' months', '').str.strip().astype(int)
df['int_rate'] = df['int_rate'].astype(str).str.replace('%', '').str.strip().astype(float)
df['revol_util'] = df['revol_util'].astype(str).str.replace('%', '').str.strip().astype(float)
df['zip_code'] = df['zip_code'].astype(str).str.replace('xx', '').str.strip().astype(int)

df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,...,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,open_acc,pub_rec,revol_bal,revol_util,total_acc,pub_rec_bankruptcies
0,5000,5000,4975.0,36,10.65,162.87,B,B2,10+,RENT,...,0,1985-01-01,1,0.0,3,0,13648,83.7,9,0
1,2500,2500,2500.0,60,15.27,59.83,C,C4,< 1,RENT,...,0,1999-04-01,5,0.0,3,0,1687,9.4,4,0
2,2400,2400,2400.0,36,15.96,84.33,C,C5,10+,RENT,...,0,2001-11-01,2,0.0,2,0,2956,98.5,10,0
3,10000,10000,10000.0,36,13.49,339.31,C,C1,10+,RENT,...,0,1996-02-01,1,35.0,10,0,5598,21.0,37,0
4,3000,3000,3000.0,60,12.69,67.79,B,B5,1,RENT,...,0,1996-01-01,0,38.0,15,0,27783,53.9,38,0


### Clean 'emp_length' column to remove text and convert to numeric

In [96]:
df['emp_length'] = df['emp_length'].astype(str).str.replace(' years', '').str.replace(' year', '').str.strip()
df['emp_length'] = df['emp_length'].replace({'< 1': '0', '10+': '10'}).astype(int)

df['emp_length'].head()

0    10
1     0
2    10
3    10
4     1
Name: emp_length, dtype: int64

### Verify if there are no more missing values

In [ ]:
missingColumnValues = df.isnull().sum()
missingColumnValues

loan_amnt                 0
funded_amnt               0
funded_amnt_inv           0
term                      0
int_rate                  0
installment               0
grade                     0
sub_grade                 0
emp_length                0
home_ownership            0
annual_inc                0
verification_status       0
issue_d                   0
loan_status               0
purpose                   0
zip_code                  0
addr_state                0
dti                       0
delinq_2yrs               0
earliest_cr_line          0
inq_last_6mths            0
mths_since_last_delinq    0
open_acc                  0
pub_rec                   0
revol_bal                 0
revol_util                0
total_acc                 0
pub_rec_bankruptcies      0
dtype: int64

### Standardize precision for float columns

In [86]:
# Get data types of all columns to standardize precision for floats
df.dtypes

loan_amnt                          int64
funded_amnt                        int64
funded_amnt_inv                  float64
term                               int64
int_rate                         float64
installment                      float64
grade                             object
sub_grade                         object
emp_length                        object
home_ownership                    object
annual_inc                       float64
verification_status               object
issue_d                   datetime64[ns]
loan_status                       object
purpose                           object
zip_code                          object
addr_state                        object
dti                              float64
delinq_2yrs                        int64
earliest_cr_line          datetime64[ns]
inq_last_6mths                     int64
mths_since_last_delinq           float64
open_acc                           int64
pub_rec                            int64
revol_bal       

In [89]:
# Standardize precision for float columns
df['funded_amnt_inv'] = df['funded_amnt_inv'].round(2)
df['int_rate'] = df['int_rate'].round(2)
df['installment'] = df['installment'].round(2)
df['annual_inc'] = df['annual_inc'].round(2)
df['dti'] = df['dti'].round(2)
df['mths_since_last_delinq'] = df['mths_since_last_delinq'].round(2)
df['revol_util'] = df['revol_util'].round(2)

df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,...,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,open_acc,pub_rec,revol_bal,revol_util,total_acc,pub_rec_bankruptcies
0,5000,5000,4975.0,36,10.65,162.87,B,B2,10+,RENT,...,0,1985-01-01,1,0.0,3,0,13648,83.7,9,0
1,2500,2500,2500.0,60,15.27,59.83,C,C4,< 1,RENT,...,0,1999-04-01,5,0.0,3,0,1687,9.4,4,0
2,2400,2400,2400.0,36,15.96,84.33,C,C5,10+,RENT,...,0,2001-11-01,2,0.0,2,0,2956,98.5,10,0
3,10000,10000,10000.0,36,13.49,339.31,C,C1,10+,RENT,...,0,1996-02-01,1,35.0,10,0,5598,21.0,37,0
4,3000,3000,3000.0,60,12.69,67.79,B,B5,1,RENT,...,0,1996-01-01,0,38.0,15,0,27783,53.9,38,0


### Standardize text columns to lowercase

In [58]:
# Standardize text columns to lowercase
df['home_ownership'] = df['home_ownership'].str.lower()
df['verification_status'] = df['verification_status'].str.lower()
df['purpose'] = df['purpose'].str.lower()
df['loan_status'] = df['loan_status'].str.lower()
df['grade'] = df['grade'].str.lower()
df['sub_grade'] = df['sub_grade'].str.lower()
df['emp_length'] = df['emp_length'].str.lower()
df['addr_state'] = df['addr_state'].str.lower()

df[['home_ownership', 'verification_status', 'purpose', 'loan_status', 'grade', 'sub_grade', 'emp_length', 'addr_state']].head()

,home_ownership,verification_status,purpose,loan_status,grade,sub_grade,emp_length,addr_state
0,rent,verified,credit_card,fully paid,b,b2,10+,az
1,rent,source verified,car,charged off,c,c4,< 1,ga
2,rent,not verified,small_business,fully paid,c,c5,10+,il
3,rent,source verified,other,fully paid,c,c1,10+,ca
4,rent,source verified,other,current,b,b5,1,or


## Invalid Values
- Date columns are formatted to date format.
- Column 'pub_rec_bankruptcies' is converted from float to int since it does not have float values.

### Convert date columns to datetime type

In [ ]:
df['issue_d'] = pd.to_datetime(df['issue_d'], format='%b-%y')
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], format='%b-%y')

df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,...,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,open_acc,pub_rec,revol_bal,revol_util,total_acc,pub_rec_bankruptcies
0,5000,5000,4975.0,36,10.65,162.87,B,B2,10+,RENT,...,0,1985-01-01,1,0.0,3,0,13648,83.7,9,0
1,2500,2500,2500.0,60,15.27,59.83,C,C4,< 1,RENT,...,0,1999-04-01,5,0.0,3,0,1687,9.4,4,0
2,2400,2400,2400.0,36,15.96,84.33,C,C5,10+,RENT,...,0,2001-11-01,2,0.0,2,0,2956,98.5,10,0
3,10000,10000,10000.0,36,13.49,339.31,C,C1,10+,RENT,...,0,1996-02-01,1,35.0,10,0,5598,21.0,37,0
4,3000,3000,3000.0,60,12.69,67.79,B,B5,1,RENT,...,0,1996-01-01,0,38.0,15,0,27783,53.9,38,0


### Convert 'pub_rec_bankruptcies' to integer type since it does not have float values

In [ ]:
print(df['pub_rec_bankruptcies'].unique())
print(df['pub_rec_bankruptcies'].isnull().sum())

df['pub_rec_bankruptcies'] = df['pub_rec_bankruptcies'].astype(int)
df['pub_rec_bankruptcies'].dtypes

[0 1 2]
0


dtype('int64')

In [95]:
df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_length,home_ownership,...,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,open_acc,pub_rec,revol_bal,revol_util,total_acc,pub_rec_bankruptcies
0,5000,5000,4975.0,36,10.65,162.87,B,B2,10+,RENT,...,0,1985-01-01,1,0.0,3,0,13648,83.7,9,0
1,2500,2500,2500.0,60,15.27,59.83,C,C4,< 1,RENT,...,0,1999-04-01,5,0.0,3,0,1687,9.4,4,0
2,2400,2400,2400.0,36,15.96,84.33,C,C5,10+,RENT,...,0,2001-11-01,2,0.0,2,0,2956,98.5,10,0
3,10000,10000,10000.0,36,13.49,339.31,C,C1,10+,RENT,...,0,1996-02-01,1,35.0,10,0,5598,21.0,37,0
4,3000,3000,3000.0,60,12.69,67.79,B,B5,1,RENT,...,0,1996-01-01,0,38.0,15,0,27783,53.9,38,0


# Cleansed Data Columns

| Column Name             | Why Selected / Rationale                                                                                  |
|-------------------------|-----------------------------------------------------------------------------------------------------------|
| loan_amnt               | Original loan amount                                                                                      |
| funded_amnt             | Amount actually funded                                                                                    |
| funded_amnt_inv         | Amount funded by investors - relevant for investor analysis                                               |
| **term**                | Loan term (months); affects risk and payment schedule.                                                    |
| int_rate                | Interest rate; primary driver of loan cost and risk.                                                      |
| installment             | Monthly payment amount; relates to borrower's payment burden.                                             |
| **grade**               | Lending Club's assigned loan grade; summarizes credit risk.                                               |
| sub_grade               | More granular risk segmentation within grade.                                                             |
| emp_length              | Borrower's employment length; relates to job stability.                                                   |
| home_ownership          | Indicates borrower's housing situation; relates to stability and risk.                                    |
| annual_inc              | Borrower's annual income; related to affordability and risk assessment.                                   |
| verification_status     | Whether income was verified; impacts confidence in reported income.                                       |
| issue_d                 | Loan issue date; useful for time-based analysis.                                                          |
| loan_status             | Target variable: current status of the loan (e.g., fully paid, charged off, current).                     |
| **purpose**             | Reason for the loan; helps segment risk by loan purpose.                                                  |
| zip_code                | Partial zip code; can be used for geographic analysis.                                       |
| addr_state              | State of residence; useful for regional risk analysis.                                                |
| **dti**                 | Debt-to-income ratio; key risk metric.                                                                    |
| **delinq_2yrs**         | Number of delinquencies in past 2 years; historical credit risk indicator.                                |
| earliest_cr_line        | Date of earliest credit line; proxy for credit history length.                                            |
| inq_last_6mths          | Number of recent credit inquiries; higher values may indicate risk.                                       |
| mths_since_last_delinq  | Months since last delinquency; recency of credit problems.                                                |
| open_acc                | Number of open credit lines; credit utilization context.                                                  |
| pub_rec                 | Number of derogatory public records; strong risk indicator.                                               |
| revol_bal               | Revolving balance (e.g., credit cards); indicates current debt load.                                      |
| **revol_util**          | Revolving line utilization rate; high values may indicate risk.                                           |
| total_acc               | Total number of credit lines; context for credit experience.                                              |
| pub_rec_bankruptcies    | Number of public record bankruptcies; strong negative credit event.                                       |